packages/lib/src/vis/shared/SelectionTool.tsx

```diff
@@ -26,78 +28,96 @@ function SelectionTool(props: Props) {
     modifierKey,
   } = props;
 
-  const [startPoint, setStartPoint] = useState<Vector2>();
-  const [endPoint, setEndPoint] = useState<Vector2>();
-  const [isAllowed, toggleAllowed] = useToggle(!modifierKey);
   const camera = useThree((state) => state.camera);
-
   const { worldToData } = useAxisSystemContext();
+
+  const [startPoint, setStartPoint] = useState<Vector2>();
+  const [endPoint, setEndPoint] = useRafState<Vector2 | undefined>(undefined);
+  const [isVisible, toggleVisible] = useToggle(true);
+
   const onPointerDown = useCallback(
     (evt: CanvasEvent<PointerEvent>) => {
-      if (!isAllowed) {
+      const { unprojectedPoint, sourceEvent } = evt;
+      if (!checkModifierKey(modifierKey, sourceEvent)) {
         return;
       }
 
-      const { unprojectedPoint, sourceEvent } = evt;
       const { target, pointerId } = sourceEvent;
       (target as Element).setPointerCapture(pointerId);
-
       setStartPoint(worldToData(unprojectedPoint));
-      setEndPoint(undefined);
-      onSelectionStart?.();
+
+      if (onSelectionStart) {
+        onSelectionStart();
+      }
     },
-    [isAllowed, worldToData, onSelectionStart]
+    [modifierKey, onSelectionStart, worldToData]
   );
 
   const onPointerMove = useCallback(
     (evt: CanvasEvent<PointerEvent>) => {
       if (!startPoint) {
         return;
       }
-      const point = worldToData(boundPointToFOV(evt.unprojectedPoint, camera));
+
+      const { unprojectedPoint, sourceEvent } = evt;
+      const point = worldToData(boundPointToFOV(unprojectedPoint, camera));
       setEndPoint(point);
 
-      if (isAllowed && onSelectionChange) {
+      if (onSelectionChange && checkModifierKey(modifierKey, sourceEvent)) {
```

Comment:
It made more sense to me to have the `checkModifierKey` check before but it does not matter.

Reply:
Well, it's less work to do if `onSelectionChange` is not passed (which is likely).

Reply:
True :+1: